Add update_all method

aropie opened the pull request:
### Overview
Currently, the only way for changes to be reflected in the Autocompleter is to `remove_all`, then `store_all` and `remove_cache`. This makes for a slow and inefficient process because most of the objects weren't really updated. 

This PR introduces the method `update_all`, which in turn calls `update_provider` for each of the Autocompleter's provider. This method aims to significantly reduce the number of operations made on Redis by bulking them together as much as possible, as well as analyze the data in order to only deal with objects which were actually modified, leaving the rest untouched.
In order to do this, one of the main points was to add a new Redis hashmap `djac.provider.sc` which stores a mapping of obj_id -> score. This is done in order to have a reference to know when a given object has updated its score, since all objects use the same score across all the ranked sets.

I've tried documenting the method and the reasoning as much as possible, but I agree the whole implementation still feels bulky and somewhat convoluted. Any suggestions on making it leaner/more concise are very welcome

### How to test
- [ ] New test cases make sense and pass

anthonyp97 reviewed (commented):
This is awesome work !!! 🔥 🧠 

Since this logic relies on the new score mapping you made, I assume we need to allow these sets get populated when we initialize autocompleters and after that we will start to see the benefits of this work?

Left a few comments and suggestions, your variable naming pattern made this much easier to read 🙏

aropie commented on autocompleter/base.py:
```diff
@@ -1071,3 +1119,300 @@ def normalize_rounding(value):
+        scores_map_key = SCORE_MAP_BASE_NAME % provider_name
+        scores_db_map = {}
+        for obj_id, score in REDIS.hgetall(scores_map_key).items():
+            parsed_score = self._deserialize_data(score) if score != b"inf" else "inf"
```
It's to go around `inf` scores stored in Redis. Since `inf` is not supported by the JSON spec, deserializing it throws an error. If the stored score is `b'inf'`, then we leave it as `'inf'` since a couple of lines later we'll transform it into a `float`. Left a comment as well in the code explaining this

aropie commented on autocompleter/base.py:
```diff
@@ -1071,3 +1119,300 @@ def normalize_rounding(value):
+        def _facet_list_to_set(facet_list):
+            return frozenset((f["key"], f["value"]) for f in facet_list)
+
+        def _get_prefixes_set(norm_terms_list):
```
Extracted inner function to a static method in `AutocompleterBase`.

aropie commented on autocompleter/base.py:
```diff
@@ -1071,3 +1119,300 @@ def normalize_rounding(value):
+        }
+        # Since we're comparing sets whose elements have the form (obj_id, {set of prefixes}), then any
+        # difference between old and new terms will show up in the symmetric difference
+        objs_with_updated_prefixes = {
```
That is absolutely true! The prefixes logic was the one that I worked first since, so when I went back to exact terms I didn't notice that I could reuse the same dataset. Merged prefixes and exact terms sections into a single one

aropie commented:
- Fixed typos
- Renamed Redis score map to `provider.sm`
- Gave more succint names to the Redis keys variables
- Added comment explaining the special treatment when we have `score=inf`
- Extracted inner method to get set of prefixes into a a static method and replaced the logic in `store` to use it
- Moved chunks of the method around trying to organize it better, including merging exact terms and prefixes logic into a single section per your suggestion.

> Since this logic relies on the new score mapping you made, I assume we need to allow these sets get populated when we initialize autocompleters and after that we will start to see the benefits of this work?
That's correct. We'd need to do one final "old way" of updating and from then on we can `update_all()` all the way

anthonyp97 reviewed (approved):
Awesome I tried manually running the update all and then doing the old remove + store method and then the next time I did update all it made no updates 🥳 .

One thing that would be nice is if we could add logging we can see in datadog that shows the counts of updates being made on these redis sets. Just so there's an easy way for us to see that this is working properly / how many updates we generally do when re initializing calc autocompleters. You'll have to look into the easiest way to get the logging to send to datadog, in general it'll require the logs being written to a file `/sites/ycharts/data/logs/management_commands/systems_initialize_autocompleters.log` which we set up in our base command logic in ycharts repo. Let me know if you need help.
